### PR TITLE
 Refactoring configuration tool 

### PR DESCRIPTION
**Pull request by AsuraZeng**
Add multithreaded execution
logic placed in the background without blocking the UI

Signed-off-by: chao zeng <chao.zeng@siemens.com>

### CODE REVIEW

File: recipes-app/switchserialmode/files/src/switchserialmode.c [hunk 1]
```diff
@@ -934,7 +934,7 @@ static E_CMD_EXE_STATUS cp210x_cmd_print_mode(int argc, char **argv)
     }
     else
     {
-        printf("gpio reset-logic(%s)\n", GET_CP2102N_GPIO2_RESET_LATH(config) ? "high" : "low");
+        printf("(%s)\n", GET_CP2102N_GPIO2_RESET_LATH(config) ? "rs422" : "rs232");
```
Comment (jan-kiszka) [inline recipes-app/switchserialmode/files/src/switchserialmode.c:937]: Unrelated change?

**Comment by AsuraZeng** [comment]
Currently in self-testing

**Comment by jan-kiszka** [comment]
Are these two changes in one commit, or is the serial configuration really that tightly coupled to threading?

**Comment by jan-kiszka** [comment]
BTW, the CI failure looks like an unrelated github connectivity issue. Maybe retry - e.g. when you have a split-up series.

**Comment by AsuraZeng** [comment]
> Are these two changes in one commit, or is the serial configuration really that tightly coupled to threading?

yes. I have sepreated this into two commits. pls help to review.

**Comment by AsuraZeng** [comment]
@jan-kiszka  I tested it and currently found no problems. Do you have any comments?

**Review by jan-kiszka — APPROVED** [review]

**Comment by BaochengSu** [comment]
This patch is planed to merge into the master after V1.3.1 release.

**Comment by BaochengSu** [comment]
@AsuraZeng Please rebase on the latest master.

**Comment by AsuraZeng** [comment]
transfer to draft. still needs some refines

File: recipes-core/customizations-example/files/board-configuration.json [hunk 2]
```diff
@@ -1,28 +1,32 @@
 {
-    "User_configuration": {
+    "External_serial": {
```
Comment (BaochengSu) [inline recipes-core/customizations-example/files/board-configuration.json:2]: What does `External Serial` mean? the DB9?

The word `external` sounds vague to me, is the arduino uart serial external or internal?

Comment (AsuraZeng) [reply recipes-core/customizations-example/files/board-configuration.json:2]: > What does `External Serial` mean? the DB9?
> 
> The word `external` sounds vague to me, is the arduino uart serial external or internal?

at here, it point to the DB9 not to the arduino.

**Comment by AsuraZeng** [comment]
@jan-kiszka @BaochengSu  hi, do there any new comments regarding this topic?

**Comment by BaochengSu** [comment]
```
iot2050setup: Abstract serval class to management different utility
Add different class to management different utility.
```

Fix the typo and grammar error of the commit message.

File: recipes-app/iot2050setup/files/iot2050setup.py [hunk 3]
```diff
@@ -275,6 +255,177 @@ def getPullMode(self, index):
...
+        while True:
+            if self.topmenu.boardType.startswith('IOT2050 Basic'):
+                modeItems = [('RS232', self.configureBasicRs232SerialMode),
+                                ('RS485', self.configureBasicRs485SerialMode),
+                                ('RS422', self.configureBasicRs422SerialMode)]
+            elif self.topmenu.boardType.startswith('IOT2050 Advanced'):
+                modeItems = [('RS232', self.configureAdvancedRs232SerialMode),
+                                ('RS485', self.configureAdvancedRs485SerialMode),
+                                ('RS422', self.configureAdvancedRs422SerialMode)]
+
```
Comment (BaochengSu) [inline recipes-app/iot2050setup/files/iot2050setup.py:280]: Indentation.

File: recipes-app/iot2050setup/files/iot2050setup.py [hunk 4]
```diff
@@ -275,6 +255,177 @@ def getPullMode(self, index):
...
+            setuptime = str(int(setuptime, 16))
+            holdtime = str(int(holdtime, 16))
+            command += ' -s ' + setuptime + ' -o ' +  holdtime
+        subprocess.call(command, shell=True)
+
+        if self.topmenu.boardType == 'IOT2050 Advanced':
+            ButtonChoiceWindow(screen=self.topmenu.gscreen,
+                            title='Note',
+                            text='You need to power cycle the device for the changes to take effect',
+                            buttons=['Ok'])
```
Comment (BaochengSu) [inline recipes-app/iot2050setup/files/iot2050setup.py:318]: Indentation.

File: recipes-app/iot2050setup/files/iot2050setup.py [hunk 5]
```diff
@@ -275,6 +255,177 @@ def getPullMode(self, index):
         elif self.checkPullModeConfig('Hiz', index):
             return 'Hiz'
 
+
+class ExternalSerialMode(BoardConfigurationUtility):
+    def __init__(self, topmenu):
+        self.topmenu = topmenu
+        super().__init__()
```
Comment (BaochengSu) [inline recipes-app/iot2050setup/files/iot2050setup.py:268]: If topmenu is shared then put it into the base class.

File: recipes-app/iot2050setup/files/iot2050setup.py [hunk 6]
```diff
@@ -493,153 +644,37 @@ def show_arduino_pinout(self):
...
-            for i in range(0, 20):
-                ioInfor += ' IO{:<3}'.format(str(i))
-                ioInfor += '| {:<11}'.format(self.config['User_configuration']['IO' + str(i) + '_MODE'])
-                ioInfor += '| ' + ' | '.join(self.config['Arduino_pinmux_map']['IO' + str(i)])
-                ioInfor += '\n'
+            m2Info = "Option | B-KEY | E-KEY | Recommend"
+            m2Capabililty = [('1    |       AutoDetect ',self.m2AutoDetect),
+                                ('2    | USB3.0 | PCIE | 5G WIFI/BT', self.m2_select_usb30_pcie),
+                                ('3    | PCIE   | PCIE | 5G WIFI/BT', self.m2_select_pcie_pcie),
+                                ('4    | PCIEx2 | ---- | SSD',        self.m2_select_pciex2)]
```
Comment (BaochengSu) [inline recipes-app/iot2050setup/files/iot2050setup.py:694]: Indentation.

File: recipes-app/iot2050setup/files/iot2050setup.py [hunk 7]
```diff
@@ -667,29 +702,27 @@ def m2_select_pcie_pcie(self):
...
-                                items=m2Capabililty,
-                                buttons=[('Ok', 'ok'),('Back', 'back', 'ESC')],
-                                default=self.currentM2Select())
+class PeripheralsMenu:
+    def __init__(self, topmenu):
+        self.topmenu = topmenu
 
+    def show(self):
+        menuItems=[('Configure External COM Ports', ExternalSerialMode(self.topmenu)),
+                    ('Configure Arduino I/O', ArduinoIoMode(self.topmenu))]
```
Comment (BaochengSu) [inline recipes-app/iot2050setup/files/iot2050setup.py:745]: Indentation

File: recipes-app/iot2050setup/files/iot2050setup.py [hunk 8]
```diff
@@ -18,6 +18,21 @@
...
+    def wrapper(*args, **kwargs):
+        thread = threading.Thread(target=func, args=args, kwargs=kwargs)
+        thread.setDaemon(True)
+        thread.start()
+        thread.join(3)
+        if thread.is_alive():
+            ButtonChoiceWindow(screen=TopMenu().gscreen,
+            title='Warnning',
+            text='Configuration may not take effect',
+            buttons=['Ok'])
```
Comment (BaochengSu) [inline recipes-app/iot2050setup/files/iot2050setup.py:33]: Indentation

File: recipes-core/customizations-example/files/board-configuration.json (hunk 2 above)
Comment (BaochengSu) [reply recipes-core/customizations-example/files/board-configuration.json:2]: Then better rename it to something like `DB9_connector`

File: recipes-app/iot2050setup/files/iot2050setup.py [hunk 9]
```diff
@@ -548,15 +539,15 @@ def configureArduinoSpi(self):
     def arduinoSpiChoice(self, select):
         if select == 'enable':
             spi = mraa.Spi(0)
-            self.setPinmuxOfUserConfig('SPI')
-        elif select == 'disable' and self.checkPinmuxConfig('SPI'):
-            self.resetPinmuxOfUserConfig('SPI')
+            self.setArduinoPinmuxCfg('SPI')
+        elif select == 'disable' and self.checkPinmuxConfig(('SPI_SS','SPI_MOSI','SPI_MISO','SPI_SCK'), (10, 11, 12, 13)):
```
Comment (BaochengSu) [inline recipes-app/iot2050setup/files/iot2050setup.py:543]: Use space between string. style.

File: recipes-core/customizations-example/files/board-configuration.json (hunk 2 above)
Comment (AsuraZeng) [reply recipes-core/customizations-example/files/board-configuration.json]: ok, change the name to `External_db9_serial`

File: recipes-app/iot2050setup/files/iot2050setup.py (hunk 5 above)
Comment (AsuraZeng) [reply recipes-app/iot2050setup/files/iot2050setup.py:269]: this is not from base class

**Review by BaochengSu — APPROVED** [review]

**Comment by BaochengSu** [comment]
Does the version number printable on device for both the swithserial and iot2050setup?

**Comment by AsuraZeng** [comment]
> Does the version number printable on device for both the swithserial and iot2050setup?

It cannot be printed at this time. could use `dpkg` to show deb package version

**Comment by AsuraZeng** [comment]
@jan-kiszka  hi, jan. do you have any comments regarding this PR?

**Comment by jan-kiszka** [comment]
Maybe just one refactoring suggestion: As patch 6 breaks bisection, and it takes patch 7 changing another tool, why not driving both the generation of the board config as well as the application out of iot2050setup modules? Maybe add a headless mode of iot2050setup to apply the config?

**Comment by AsuraZeng** [comment]
> Maybe just one refactoring suggestion: As patch 6 breaks bisection, and it takes patch 7 changing another tool, why not driving both the generation of the board config as well as the application out of iot2050setup modules? Maybe add a headless mode of iot2050setup to apply the config?

@jan-kiszka  hi,jan. Have already modified it, pls help to review it. thanks a lot.

File: recipes-core/customizations-example/customizations-example_0.1-iot2050-debian.bb [hunk 10]
```diff
@@ -19,9 +19,7 @@ DEBIAN_DEPENDS = "openssh-server, bluez, pulseaudio-module-bluetooth, network-ma
 SRC_URI = " \
     file://status-led.service \
     file://postinst \
-    file://board-configuration \
     file://board-configuration.service \
-    file://board-configuration.json \
```
Comment (jan-kiszka) [inline recipes-core/customizations-example/customizations-example_0.1-iot2050-debian.bb:24]: Looks inconsistent to me: board-configuration tool is factored out, but the file describing it as service is no? And no dependency is modeled between that package and this customization thing here?

Comment (AsuraZeng) [reply recipes-core/customizations-example/customizations-example_0.1-iot2050-debian.bb:24]: > Looks inconsistent to me: board-configuration tool is factored out, but the file describing it as service is no? And no dependency is modeled between that package and this customization thing here?

Put related resources together, pls review.

**Review by jan-kiszka — APPROVED** [review]
LGTM
